lingua_franca/lang/parse_pt.py

```diff
@@ -14,26 +14,66 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-"""
-    Parse functions for Portuguese (PT-PT)
-
-    TODO: numbers greater than 999999
-    TODO: date time pt
-"""
 
 from datetime import datetime
 from dateutil.relativedelta import relativedelta
-from lingua_franca.lang.parse_common import is_numeric, look_for_fractions
-from lingua_franca.lang.common_data_pt import _FRACTION_STRING_PT, \
-    _PT_ARTICLES, _PT_NUMBERS
 
+from lingua_franca.lang.parse_common import is_numeric, look_for_fractions, \
+    invert_dict, ReplaceableNumber, partition_list, tokenize, Token, extract_numbers_generic
+
+from lingua_franca.lang.common_data_pt import _PT_ARTICLES, _NUM_STRING_PT, \
+    _LONG_ORDINAL_PT, _LONG_SCALE_PT, _SHORT_SCALE_PT, _SHORT_ORDINAL_PT, \
+    _FRACTION_MARKER_PT, _DECIMAL_MARKER_PT, _PT_NUMBERS, _SUFFIX_FRACTION_MARKER_PT, \
+    _NEGATIVES_PT, _NEGATIVE_SUFFIX_MARKER_PT, _SUM_MARKER_PT,\
+    _FEMALE_ENDINGS_PT, _MALE_ENDINGS_PT, _MALE_DETERMINANTS_PT, _FEMALE_DETERMINANTS_PT, _GENDERS_PT
+
+from lingua_franca.lang.format_pt import pronounce_number_pt
+
+
+def generate_plurals_pt(originals):
+    """
+    Return a new set or dict containing the plural form of the original values,
+
+    In Portuguese this means all with 's' appended to them.
+
+    Args:
+        originals set(str) or dict(str, any): values to pluralize
+
+    Returns:
+        set(str) or dict(str, any)
+
+    """
+    if isinstance(originals, dict):
+        return {key + 's': value for key, value in originals.items()}
+    return {value + "s" for value in originals}
+
+
+_MULTIPLIES_LONG_SCALE_PT = set(_LONG_SCALE_PT.values()) | \
+                            generate_plurals_pt(_LONG_SCALE_PT.values())
+
+_MULTIPLIES_SHORT_SCALE_PT = set(_SHORT_SCALE_PT.values()) | \
+                             generate_plurals_pt(_SHORT_SCALE_PT.values())
+
+_STRING_NUM_PT = invert_dict(_NUM_STRING_PT)
+_STRING_NUM_PT.update(generate_plurals_pt(_STRING_NUM_PT))
+_STRING_NUM_PT.update({
+    "meio": 0.5,
+    "meios": 0.5,
+    "par": 2
```

Comment:
I was also wondering about this for Spanish. Should the text two pairs return 4? If so, should eight dozens return 96? Isn't a pair some sort of object also? For instance, in a hand of poker, a pair should not be translated to 2. What about words like a couple? How do we handle this?

Reply:
the reasoning here was that for english "a couple of X" makes sense to be extracted since it is used in everyday language, ("hey mycroft grab a couple of beers")

but in portuguese/spanish there isnt a clear distinction between "pair" and "couple", "a pair of jeans" would not make sense to extract number 2, but "a couple of jeans" would. 

I think we should not consider "pair" == 2, i will remove that from this PR

Reply:
in regards to "should eight dozens return 96", i dont think so, this would make sense in [extract units](https://github.com/MycroftAI/lingua-franca/pull/2) since that takes unit dimensions into account

performing math is out of scope for this utility method

Reply:
Thank you for those clarifications, I'll bear them in mind.

Reply:
I think 'pair' and 'couple' are translated into spanish like 'par' and 'pareja'. The former means '2' while the later is used to group things and people in two.

So, I wouldn't remove "pair" == 2...

Reply:
let's revisit this topic later, for now i will remove it since i can't think of any relevant example in portuguese where i want to extract it

i left it commented out with a comment pointing to this discussion